Add support for Alpine on LGBM

Renato2000 opened the pull request:
This commit makes it possible to use LGBM on nodes with AMD64 architecture using Alpine

Commit:
Add support for Alpine on LGBM

This commit makes it possible to use LGBM on nodes with AMD64 architecture using Alpine

artpdr commented on make.sh:
```diff
@@ -76,9 +76,11 @@ if [[ "$ARCH_BUILD" != "single" ]]; then
 echo_stage "Launching AMD64 container..."
 container_amd64=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env-amd64)
-docker cp docker/make_lightgbm.sh $container_amd64:/lightgbm
+docker cp docker/make_lightgbm.sh $container_amd64:/lightgbm/make_lightgbm.sh
```
Why is this change necessary?

Renato2000 replied:
That line was giving errors on the Alpine version so I also changed it for amd64. However it was actually not necessary and so I reverted it in the last commit.

artpdr commented on make.sh:
```diff
@@ -110,6 +121,12 @@ if [[ "$ARCH_BUILD" != "single" ]]; then
 fi
 
+FROM_ALPINE="$container_alpine:/lightgbm/LightGBM"
+docker cp $FROM_ALPINE/lib_lightgbm.so build/alpine
```
The README needs to be updated to account for this new build/alpine. And probably this new build/alpine should be inside the amd64 folder. Please check with other members the best structure having in mind that changing the directory structure may be a breaking change and will require a major release.

Renato2000 replied:
I moved the alpine build into the amd64 folder. So, the new structure is: build/amd64/glibc, build/amd64/alpine and build/arm64. I will also update the README in the next commit.

Commit:
Change build for LGBM

This commit changes the build folder for LGBM, moving the Alpine build into amd64.

Commit:
Update README with the new build structure

gandola commented on docker/lightgbm-ci-build-env-alpine/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.18.4
+
+RUN apk add clang git cmake make gcc g++ tar zip curl openjdk8 bash
```
is openmp being installed and used?

Renato2000 replied:
I think openmp in being installed alongside g++. Since it also installs gcc v12.2.1 and according to this [gcc.gnu.org/wiki/openmp](https://gcc.gnu.org/wiki/openmp), it fully supports OpenMP 4.5.

gandola commented on README.md:
```diff
@@ -43,9 +43,14 @@ This is the output:
+│   │   ├── libgomp.so.1.0.0
+│   │   ├── lib_lightgbm.so
+│   │   └── lib_lightgbm_swig.so
+│   ├── alpine
```
don't make it `alpine`, `musl` is the correct term here.

Renato2000 replied:
I changed it to musl in the last commit.

gandola commented on docker/make_docker_image.sh:
```diff
@@ -29,6 +29,10 @@ case "$1" in
         docker build -t lightgbm-ci-build-env-amd64 lightgbm-ci-build-env-amd64
         ;;
+    alpine)
+        echo "Building docker image for Alpine"
```
Alpine with Musl

gandola commented on docker/lightgbm-ci-build-env-alpine/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.18.4
```
What are the software versions on this alpine version? clang, cmake, etc? 

Lets use Zulu JDK which is the one being used by other builds.

Renato2000 replied:
The software versions are:
  
  cmake=3.26.5
  make=4.4.1
  gcc  and g++: 12.2.1
  
  In the last commit I specified the version when running apk add. Also, clang was removed because it was not necessary.

Renato2000 replied:
I also installed the same version of Zulu JDK that is being used on the other builds.

Commit:
Add versions on Alpine dockerfile

This commit adds versions for java, cmake, gcc and g++.

Commit:
Change make_docker_image image creation message for Alpine

Commit:
Change build structure for Alpine

Instead of storing the build made for Alpine with musl on amd64/alpine, it is now located in amd64/musl.

Commit:
Fix Alpine dockerfile

Some tests tests where accidently added in the latest commit, which are removed.

Commit:
Remove version from make on Alpine dockerfile

apk is not able to find the specified version, so it was removed.

gandola commented:
LGTM!
